internal/config/cgroupmanager/supported.go

```diff
@@ -0,0 +1,123 @@
+// +build linux
+
+package cgroupmanager
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+
+	libpodcgroups "github.com/containers/libpod/pkg/cgroups"
+	libctrcgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/pkg/errors"
+)
+
+var (
+	_cgroupHasHugetlbOnce sync.Once
+	_cgroupHasHugetlb     bool
+	_cgroupHasHugetlbErr  error
+
+	_cgroupHasPidOnce sync.Once
+	_cgroupHasPid     bool
+	_cgroupHasPidErr  error
+
+	_cgroupHasMemorySwapOnce sync.Once
+	_cgroupHasMemorySwap     bool
+	_cgroupHasMemorySwapErr  error
+
+	_cgroupIsV2Once sync.Once
+	_cgroupIsV2     bool
+	_cgroupIsV2Err  error
+)
+
+// initializeCgroups initializes all of the singleton variables
+// that store the node's cgroup configuration. Currently, we check
+// hugetlb, cgroup v1 or v2, pid and memory swap support
+// We check the error at server initialization, and if we error, shutdown
+// cri-o early, instead of when we're already trying to run containers.
+func initializeCgroups() error {
+	toInit := []struct {
+		initializer func() bool
+		err         *error
+	}{
+		{
+			initializer: CgroupHasHugetlb,
+			err:         &_cgroupHasHugetlbErr,
+		},
+		{
+			initializer: CgroupHasPid,
+			err:         &_cgroupHasPidErr,
+		},
+		{
+			initializer: CgroupIsV2,
+			err:         &_cgroupIsV2Err,
+		},
+		{
+			initializer: CgroupHasMemorySwap,
+			err:         &_cgroupHasMemorySwapErr,
+		},
+	}
+	for _, i := range toInit {
+		i.initializer()
+		if *i.err != nil {
+			return errors.Wrapf(*i.err, "failed to initialize cgroup information")
+		}
+	}
+	return nil
+}
+
+// CgroupHasHugetlb returns whether the hugetlb controller is present
+func CgroupHasHugetlb() bool {
+	_cgroupHasHugetlbOnce.Do(func() {
+		if CgroupIsV2() {
+			if _, err := ioutil.ReadDir("/sys/fs/cgroup/hugetlb"); err != nil {
+				_cgroupHasHugetlbErr = errors.Wrap(err, "readdir /sys/fs/cgroup/hugetlb")
```

Comment:
I bet `iotuils.ReadDir` already provides the directory name in the error it returns, so there's no need to wrap it.

Reply:
What ioutils.ReadDir does is opens a directory, read all the entries, calls `stat(2)` on every entry and returns all that to the caller. This is a relatively expensive operation, and is surely an overkill if all you need is to check whether the directory exists.

Calling `os.Stat()` and checking `IsDir()` is probably sufficient here.

Reply:
Looks like the logic of v1/v2 is wrong here. For v1, you need to check if `/sys/fs/cgroup/hugetlb` exists and is a directory. For v2, you need to read `/sys/fs/cgroup/cgroup.controllers` and see if it contains `hugetlb`.

Reply:
fixed as suggested